File: TNetCommon/Conversions.cs

// @Author: Arpan Jati
// @Date: Dec 2014

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TNetD
{
    public class Conversions
    {
        public static byte[] Int64ToVector(long data)
        {
            byte[] _out = new byte[8];

            for (int i = 0; i < 8; i++)
            {
                _out[i] = (byte)(((data >> (8 * i)) & 0xFF));
            }

            return _out;
        }

        public static byte[] Int16ToVector(short data)
        {
            byte[] _out = new byte[2];

            for (int i = 0; i < 2; i++)
            {
                _out[i] = (byte)(((data >> (8 * i)) & 0xFF));
            }

            return _out;
        }

        public static byte[] Int32ToVector(int data)
        {
            byte[] _out = new byte[4];

            for (int i = 0; i < 4; i++)
            {
                _out[i] = (byte)(((data >> (8 * i)) & 0xFF));
            }

            return _out;
        }

        public static byte[] FloatToVector(float data)
        {
            return BitConverter.GetBytes(data);
        }

        public static byte[] DoubleToVector(double data)
        {
            return BitConverter.GetBytes(data);
        }


        public static long VectorToInt64(byte[] data)
        {
            long result;
            result = (long)data[0];
            result |= ((long)data[1]) << 8;
            result |= ((long)data[2]) << 16;
            result |= ((long)data[3]) << 24;
            result |= ((long)data[4]) << 32;
            result |= ((long)data[5]) << 40;
            result |= ((long)data[6]) << 48;
            result |= ((long)data[7]) << 56;
            return result;
        }

        public static short VectorToInt16(byte[] data)
        {
            short result;
            result = (short)data[0];
            result |= (short)(((short)data[1]) << 8);
            return result;
        }

        public static int VectorToInt32(byte[] data)
        {
            int result;
            result = (int)data[0];
            result |= ((int)data[1]) << 8;
            result |= ((int)data[2]) << 16;
            result |= ((int)data[3]) << 24;
            return result;
        }

        public static float VectorToFloat(byte[] data)
        {
            return BitConverter.ToSingle(data, 0);
        }

        public static double VectorToDouble(byte[] data)
        {
            return BitConverter.ToDouble(data, 0);
        }
    }
}

File: TNET_UI/Sources/Conversions.cpp

/*
*  @Author: Arpan Jati
*  @Version: 1.0
*/

#include "Utils.h"
#include "Conversions.h"

vector<byte> Conversions::Int64ToVector(int64_t data)
{
	vector<byte> out = vector<byte>(8);

	for (int i = 0; i < 8; i++)
	{
		out[i] = (((data >> (8 * i)) & 0xFF));
	}

	return out;
}

vector<byte> Conversions::Int16ToVector(int16_t data)
{
	vector<byte> out = vector<byte>(2);

	for (int i = 0; i < 2; i++)
	{
		out[i] = (((data >> (8 * i)) & 0xFF));
	}

	return out;
}

vector<byte> Conversions::Int32ToVector(int32_t data)
{
	vector<byte> out = vector<byte>(4);

	for (int i = 0; i < 4; i++)
	{
		out[i] = (((data >> (8 * i)) & 0xFF));
	}

	return out;
}

vector<byte> Conversions::FloatToVector(float data)
{
	unsigned char * p = reinterpret_cast<unsigned char *>(&data);
	vector<byte> out(p, p + 4);
	return out;
}

vector<byte> Conversions::DoubleToVector(double data)
{
	unsigned char * p = reinterpret_cast<unsigned char *>(&data);
	vector<byte> out(p, p + 8);
	return out;
}


int64_t Conversions::VectorToInt64(vector<byte> data)
{
		uint64_t result;
		result = (uint64_t)data[0];
		result |= ((uint64_t)data[1]) << 8;
		result |= ((uint64_t)data[2]) << 16;
		result |= ((uint64_t)data[3]) << 24;
		result |= ((uint64_t)data[4]) << 32;
		result |= ((uint64_t)data[5]) << 40;
		result |= ((uint64_t)data[6]) << 48;
		result |= ((uint64_t)data[7]) << 56;
		return result;
}

int16_t Conversions::VectorToInt16(vector<byte> data)
{
	int16_t result;
	result = (int16_t)data[0];
	result |= ((int16_t)data[1]) << 8;
	return result;
}

int32_t Conversions::VectorToInt32(vector<byte> data)
{
	int32_t result;
	result = (int32_t)data[0];
	result |= ((int32_t)data[1]) << 8;
	result |= ((int32_t)data[2]) << 16;
	result |= ((int32_t)data[3]) << 24;
	return result;
}

float Conversions::VectorToFloat(vector<byte> data)
{
	float* floatArray = reinterpret_cast<float*>(data.data());
	return floatArray[0];
}

double Conversions::VectorToDouble(vector<byte> data)
{
	double* doubleArray = reinterpret_cast<double*>(data.data());
	return doubleArray[0];
}